pkg/apps/packagekit.js

```diff
@@ -0,0 +1,185 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+var cockpit = require("cockpit");
+var $ = require("jquery");
+
+var client = cockpit.dbus("org.freedesktop.PackageKit", { superuser: "try" });
+
+var PK_STATUS_ENUM_WAIT = 1;
+var PK_STATUS_ENUM_WAITING_FOR_LOCK = 30;
+
+var PK_FILTER_INSTALLED   = (1 << 2);
+var PK_FILTER_NEWEST      = (1 << 16);
+var PK_FILTER_ARCH        = (1 << 18);
+var PK_FILTER_NOT_SOURCE  = (1 << 21);
+
+function transaction(method, args, progress_cb, package_cb) {
+    var defer = cockpit.defer();
+
+    client.call("/org/freedesktop/PackageKit", "org.freedesktop.PackageKit", "CreateTransaction", [ ]).
+        done(function(path_result) {
+            var tr = client.proxy("org.freedesktop.PackageKit.Transaction", path_result[0]);
+            var cancelled = false;
+            var allow_wait_status = false;
+
+            function cancel() {
+                tr.Cancel();
+                cancelled = true;
+            }
+
+            function changed() {
+                if (progress_cb && defer.promise().state() == "pending") {
+                    var data = {
+                        waiting: false,
+                        percentage: 0,
+                        cancel: null
+                    };
+
+                    if (allow_wait_status &&
+                        (tr.Status == PK_STATUS_ENUM_WAIT || tr.Status == PK_STATUS_ENUM_WAITING_FOR_LOCK))
+                        data.waiting = true;
+                    if (tr.Percentage !== undefined && tr.Percentage !== 101)
+                        data.percentage = tr.Percentage;
+                    if (tr.AllowCancel)
+                        data.cancel = cancel;
+
+                    progress_cb(data);
+                }
+            }
+
+            changed();
+            $(tr).on("changed", changed);
+            window.setTimeout(function () {
+                allow_wait_status = true;
+                changed();
+            }, 1000);
+
+            $(tr).on("ErrorCode", function (event, code, details) {
+                defer.reject(details, cancelled? "cancelled" : code);
+            });
+            $(tr).on("Package", function (event, info, package_id, summary) {
+                if (package_cb && defer.promise().state() == "pending")
+                    package_cb(info, package_id, summary);
+            });
+            $(tr).on("Finished", function (event, exit, runtime) {
+                defer.resolve(exit);
+            });
+            tr.call(method, args).fail(function (error) {
+                console.log("Error", error);
+                defer.reject(error);
+            });
+        }).
+        fail(function (error) {
+            defer.reject(error);
+        });
+
+    return defer.promise();
+}
+
+function progress_reporter(base, range, callback) {
+    if (callback)
```

Comment:
Needs { } as the `then` block is multiple lines.

Reply:
Fixed.